### ISSUE TITLE
Discussion on documentation

### ISSUE THREAD

**Issue by ana-pantilie** [runtimeverification/haskell-backend#2780]
@JKTKops has brought to my attention the issue that we don't have adequate documentation. Also, multiple newcomers to the team have asked about documentation which describes how some component of the code works, like how the simplifier works, and, unfortunately we don't really have anything which gives a good overview of things.

We do have the following:
1. Each global function is documented, and we stress adding documentation where it's lacking.
2. We have the design decision documents.
3. We have a context map of the modules.

No. (1) only gives very localized understanding of the code. We would want to also show how the functions behave together.

No. (2): we also have a few problems with the state they are in now: some of the design decisions are outdated, either their implementation is no longer part of the code base or the information is partially correct. For the up-to-date ones, the relationship to the code might be somewhat obscure as well.

No. (3) is a great overview of the code, but we seem to be lacking the middle layer of actually documenting what each bounded context is in detail.

So, to summarize, the issues we need to fix are:
- (1) and (3) indicate that we are missing this "middle layer", which would glue things together.
- (2) shows that we need to clean-up the docs, and organize them in such a way that it's easy to go from design decision to implementation.

I want us to discuss this problem and propose solutions.

**Comment by ana-pantilie** [comment]
My idea would be to add a new document, similar to [K's user manual](https://github.com/kframework/k/blob/master/USER_MANUAL.md), but aimed at developers. This "Developer Manual" should describe important components in the code and their implementation. I'd prefer to not use the Wiki for this because I want changes to the manual to be reviewed.

@JKTKops suggested we should add documentation incrementally, and I think that's a very good idea.

This might be obvious, but I want to be explicit about some scenarios where we need to find out how something works:
- when someone asks a question about some behavior of the backend, and answering it needs investigation (reading the code, thinking about examples)
- when working on any type of issue where understanding how some code works is necessary

We also need to know where it would be appropriate to document something, so I think we can split things like this:
1. If the logic is local to some small component (function), we need to make sure it's documented as code (meaning that the code is readable and self-documenting) and that the comment above is detailed enough to give a good understanding of what the function is intended to do. The comment should also point to any design decision documents if there are any describing the theory behind the implementation.
 2. If we learn how some functions integrate, and understand effectively "how things work" in a larger context, we should add documentation to the Developer Manual. The "upper-level places", basically where the functions integrate, should have a link to the corresponding section.

I think we can start by adding a `DEVELOPER_MANUAL.md` which will contain, at the highest level, sections corresponding to the contexts in the context map. We will need to be more in-depth than that, but we can start there and figure out how to better organize things as we go.

Thoughts?

**Comment by dopamane** [comment]
@ana-pantilie I agree and support these ideas. My only concern is that Github markdown may not be sufficient for typesetting math notation and working with a (potentially) large, growing document. In my experience, Github markdown is great for small and medium sized documents, but it becomes cumbersome to maintain for large documents with many sections, subsections, and cross-references. I suggest LaTeX as an alternative to Github markdown. 

Pros: 
 * Math notation, tables, images, bibliography, and fonts are displayed better with LaTex
 * Being able to generate a PDF or DVI etc. makes it more portable for viewing outside of Github e.g. when working offline.

Cons: 
 * We would need to automate generating a PDF (with CI?) to ensure it reflects the most up-to-date `DEVELOPER_MANUAL.tex` file
 * LaTeX is more complex than Github markdown

**Comment by andreiburdusa** [comment]
I also agree with this plan of adding documentation, and I think LaTeX is a good alternative to Github markdown

**Comment by geo2a** [comment]
I support this idea, and I agree that using LaTeX is appropriate, since we may be looking at a rather long document after a while.

I took a look again at the `docs/` folder and there are a lot of documents that seem outdated. As a newcomer, I have found [`introduction.md`](https://github.com/kframework/kore/blob/master/docs/introduction.md) to be enlightening and I think it may be a good candidate for verbatim inclusion in the Developer Manual.

I suppose we could put the Developer Manual in the `docs/` folder as well. I think we can reuse a lot of TeX definitions for notations and such from [`introduction-to-matching-logic.tex`](https://github.com/kframework/kore/blob/master/docs/introduction-to-matching-logic.tex). I think though that it is better to put the sources into a subfolder, say `docs/developer_manual`.

**Comment by nishantjr** [comment]
BTW, Pandoc allows mixing latex and markdown. It works quite well overall and is more readable than straight latex.

**Comment by ana-pantilie** [comment]
Hi @shd101wyy !

After discussing with @ehildenb during our planning meeting today, we have decided that it would be best for us to go with mixing both Markdown and LaTeX, but we're not sure if this would be feasible, so we have a couple of questions:
- Do you know if it would be possible to have a, mostly Markdown, online version of our documentation (similar to https://kframework.org/index.html), but with the added feature of adding LaTeX inline math docs?
- Would it be possible to generate a PDF of the documentation, for offline usage?

**Comment by ana-pantilie** [comment]
@shd101wyy Oh, sorry for pinging you on Slack as well, didn't see you responded so quickly! I think, ideally, we'd want a PDF matching the structure of the website. Is that right @dopamane ?

**Comment by dopamane** [comment]
@ana-pantilie @shd101wyy I was thinking of using pandoc to generate a PDF **from the markdown** thus matching its content/structure/rendering while also providing offline viewing, portability, and distribution. If possible, I think we should follow a hub-and-spoke model for deriving documentation from markdown. 

From the discussion in today's Haskell meeting, it sounds like using Github markdown is the best way to integrate documentation with the website, Kore Github project, and offline while also providing expressiveness when typesetting math notation and allowing for quick navigation. I'd like to draft a small sample document in Github markdown (and LaTeX too for comparison) demonstrating expected structure and exercising some common notation we use. Then, process it with pandoc for both PDF and HTML/Web to get a sense for the style and ergonomics.

**Comment by shd101wyy** [comment]
We could build the website first then consider to generate a PDF. I have added the support to render latex math in Markdown in pull request https://github.com/runtimeverification/k-web-theme/pull/14. Inline math needs to be written in `$...$` while block math in `$$...$$`. So please let me know when we want to start building the website. Thank you!

**Comment by ana-pantilie** [comment]
That's awesome, thank you @shd101wyy !

**Comment by ana-pantilie** [comment]
This was done. Continuing work in https://github.com/runtimeverification/haskell-backend/issues/3026
